app/models/good_job/batch.rb

```diff
@@ -86,15 +86,23 @@ def initialize(_record: nil, **properties) # rubocop:disable Lint/UnderscorePref
     # @return [Array<ActiveJob::Base>] Active jobs added to the batch
     def enqueue(active_jobs = [], **properties, &block)
       assign_properties(properties)
-      record.save!
+      if record.new_record?
+        record.save!
+      else
+        record.with_advisory_lock(function: "pg_advisory_lock") do
+          record.enqueued_at_will_change!
+          record.finished_at_will_change!
+          record.update!(enqueued_at: nil, finished_at: nil)
+        end
+      end
 
       active_jobs = add(active_jobs, &block)
 
-      record.finished_at = nil
-      record.enqueued_at = Time.current if enqueued_at.nil?
-      record.save!
+      record.with_advisory_lock(function: "pg_advisory_lock") do
+        record.update!(enqueued_at: Time.current)
+        record._continue_discard_or_finish(lock: false)
+      end
```

Comment:
Looks good! i'm agnostic about the enqueued_at reset, I think you're doing that in lieu of just putting the `add` in the advisory lock (if that's even possible)? Perfectly fine solution but i would consider enqueued_at to mean when the first jobs started coming in, doesn't matter that much though total batch statistics can be derived from the jobs.

Reply:
if `add` can raise an exception you might be able to get in a weird spot where the first callback from the first enqueue doesn't even fire

Reply:
> Perfectly fine solution but i would consider enqueued_at to mean when the first jobs started coming in...

Yes, that was my hope too but I realize it has an operational meaning of like "have the jobs been enqueued so that the finish callback should be triggerable".

>  I think you're doing that in lieu of just putting the add in the advisory lock

Exactly. I think it would get weird, like blocking any jobs from fully performing until all jobs had been inserted (right now that happens as a single Bulk insert, but I could see that not necessarily being the case in the future).

> if `add` can raise an exception you might be able to get in a weird spot where the first callback from the first enqueue doesn't even fire, and no callback is ever fired.

You're right! Hmm. I guess if the `enqueue` method had an exception raised through it, I'd expect it to be in a weird state. Hopefully I can not be bothered by that edge case for now 🤞🏻